lnst/Tests/PktGen.py

```diff
@@ -0,0 +1,185 @@
+            self._res[device] = {
+                "params": self._parse_values(params),
+                "current": self._parse_values(current),
+                "result": self._parse_results(result),
+            }
+
+    def _read_dev_outputs(self) -> Iterator[tuple[str, str]]:
+        for device in self._devs:
+            output = ""
+            with open(f"/proc/net/pktgen/{device}", "r") as f:
+                output = f.readlines()
+            yield device, "\n".join(output)
+
+    def _split_output(self, output: str) -> tuple:
+        match = re.search(
+            r"Params:(.+)Current:(.+)Result:\s(?:\w+)(?::\s(.+))", output, re.DOTALL
+        )
+        if not match:
+            raise TestModuleError("Could not parse pktgen devide output")
+
+        return match.groups()
+
+    def _parse_values(self, params) -> dict[str, str]:
+        values = {}
+
+        for key, value in re.findall(r"(\w+):?\s(\S+)", params, re.MULTILINE):
+            values[key.lower()] = value
+
+        return values
+
+    def _parse_results(self, results) -> dict[str, Any]:
+        match = re.search(
+            r"(\d+).+ usec,\s(\d+)\s\(.+,(\d+)frags\).+?.+errors:\s?(\d+)",
+            results,
+            re.DOTALL,
+        )
+        if not match:
+            raise TestModuleError("Could not parse pktgen device result")
+
+        keys = ["duration", "packets", "frags", "errors"]
+
+        try:
+            results = dict(  # converts iterator to actual dict
+                zip(  # zips values with `keys`
+                    keys, map(int, match.groups())  # map converts str numbers to int
+                )
+            )
+        except ValueError:  # just to be sure all the nums can be converted to int
+            raise TestModuleError("Could not convert numbers in pktgen results")
+
+        results["duration"] /= 1_000_000  # from usec to sec
+        return results
+
+
+class PktGen(BaseTestModule):
+    """
+    In the scope of this module, the physical interface is refered as `interface`.
+    Pktgen device (interface@anything) is refered as device.
+
+    Inspired by https://github.com/torvalds/linux/blob/master/samples/pktgen/pktgen_sample03_burst_single_flow.sh
+    """
+
+    cpus = ListParam(type=IntParam())  # each CPU is 1 generator
+
+    src_if = DeviceParam()
+    dst_mac = StrParam()
+
+    src_ip = IpParam()
+    dst_ip = IpParam()
+
+    count = IntParam(mandatory=False, default=0)  # 0 = no upper limit
+    size = IntParam(mandatory=False, default=60)  # 4 bytes are added for CRC by NIC
+    frags = IntParam(mandatory=False, default=1)
+    burst = IntParam(mandatory=False, default=8)
+
+    duration = IntParam(default=60)
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        self._devices = []
+
+        self._res_data = {}
+        self._output_parser = None
+
+    def run(self):
+        if not kmod_in_use("pktgen"):
+            raise TestModuleError("pktgen module is not loaded")
```

Comment:
Actually, we can't. The `TestModule.__init__()` method is also executed on the controller, which generally doesn't have to be the same machine as the generator (which executes the `TestModule.run()` method), so if we have 3 machines setup (controller, generator, receiver) it would crash.